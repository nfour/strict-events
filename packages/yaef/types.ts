declare type Omit<T, K extends keyof T> = T extends any ? Pick<T, Exclude<keyof T, K>> : never;
declare type TupleToUnion<T extends unknown[]> = T[number];
declare type TupleToUnionList<T extends unknown[]> = Array<T[number]>;
declare type TupleToPartialIntersection<T extends unknown[]> = UnionToPartialIntersection<TupleToUnion<T>>;
declare type TupleToIntersection<T extends unknown[]> = UnionToIntersection<TupleToUnion<T>>;
declare type TupleToIntersectionList<T extends unknown[]> = Array<TupleToIntersection<T>>;
declare type TupleToPartialIntersectionList<T extends unknown[]> = Array<TupleToPartialIntersection<T>>;
declare type UnionToIntersection<U> = (
  (U extends any
    ? (k: U) => void
    : never) extends (k: infer I) => void
      ? I
      : never
);
declare type UnionToPartialIntersection<U> = (
  (U extends any
    ? (k: U) => void
    : never) extends (k: infer I) => void
      ? Partial<I>
      : never
);
