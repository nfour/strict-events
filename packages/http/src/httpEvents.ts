import { EventSignature } from '@yaef/core';

import { IHttpMethod, IStringDict } from './types';

export const HttpRequest = EventSignature('HttpRequest', {} as {
  _eventId: string;
  body: any,
  headers: IStringDict,
  path: string,
  query: IStringDict,
  params: IStringDict,
  method: IHttpMethod,
});

export const PrepareHttpRequest = EventSignature('PrepareHttpRequest', {} as typeof HttpRequest);

export const HttpRequestResponse = EventSignature('HttpRequestResponse', {} as {
  _eventId: string;
  statusCode: number;
  body: unknown,
  headers: IStringDict,
});
