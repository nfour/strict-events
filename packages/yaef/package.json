{
  "name": "yaef",
  "version": "0.6.0",
  "main": "./build/index",
  "files": [
    "./build",
    "./src"
  ],
  "dependencies": {
    "@types/bluebird": "^3.5.26",
    "@types/debug": "^4.1.4",
    "@types/lodash": "^4.14.123",
    "bluebird": "^3.5.3",
    "debug": "^4.1.1",
    "lodash": "^4.17.13",
    "supports-color": "^6.1.0",
    "ts-node": "^8.0.3"
  },
  "devDependencies": {
    "typescript": "~3.4.5"
  },
  "engines": {
    "node": ">= 10.5.0"
  },
  "license": "MIT"
}
