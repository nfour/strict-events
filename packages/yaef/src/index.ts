export * from './mediation';
export * from './types';
export * from './componentry';
export * from './registry';
export * from './remote/remote';
export { EventAwaiter } from './lib';
