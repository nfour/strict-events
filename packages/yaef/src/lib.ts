import { config, delay } from 'bluebird';

import { IEventSignature } from './';
import { Mediator } from './mediation';

/** TODO: test me */
export function ErrorFromCallPoint ({ fromStackPosition }: { fromStackPosition: number }) {
  return (message?: string) => {
    const error = new Error(message);

    /** Add one position, because of this function call */
    const nextPosition = fromStackPosition + 1;

    const stackLines = (error.stack || '').split('\n');

    const nextStack = [
      stackLines[0],
      ...stackLines.slice(nextPosition),
    ].join('\n');

    error.stack = nextStack;

    return error;
  };
}

config({ cancellation: true });

/**
 * A factory which produces a function that can wait for an event to complete, once.
 *
 * - Set `timeout` to `Infinity` to remove the timeout error.
 *
 * The factory returns a function which when called, returns a promise with resolves with the event.
 * - Use the `filterCb` to filter an event
 *
 * @example
 *
 * const waitForEvent = EventAwait(mediator);
 * const payload = await waitForEvent(MyEvent, (payload) => payload._id === someId)
 * const payload = await waitForEvent(MySingularEvent)
 */
export function EventAwaiter<M extends Mediator<any>> (mediator: M, { timeout = 5000 }: { timeout?: number } = {}) {
  return function waitForEvent<E extends IEventSignature> (event: E, filterCb?: ((event: E) => boolean)): Promise<E> {
    return new Promise<E>((resolve, reject) => {
      /**
       * This code is probably way more verbose than necessary but whatever, find a lib later.
       */
      let isCompleted = false;

      const complete = () => {
        isCompleted = true;

        mediator.removeObserver(observeCb);
      };

      const observeCb = (payload: any) => {
        const isFilteredIn = filterCb
          ? filterCb(payload)
          : true;

        if (isCompleted || !isFilteredIn) { return; }

        complete();
        timeoutPromise.cancel();
        resolve(payload);
      };

      mediator.observe(event, observeCb);

      const timeoutPromise = delay(timeout).then(() => {
        if (isCompleted) { return; }

        complete();
        reject(new Error(`Timed out waiting for event: ${JSON.stringify(event, null, 2)}`));
      });
    });
  };
}

// TODO: add generic types
export const CheckEventId = (id: string) => ({ _eventId }: { _eventId: string }) => _eventId === id;
