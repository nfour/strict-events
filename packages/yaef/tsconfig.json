{
  "compilerOptions": {
    "strict": true,
    "incremental": true,
    "outDir": "build",
    "declaration": true,
    "tsBuildInfoFile": "./build/tsconfig.tsbuildinfo",
    "downlevelIteration": true,
    "rootDir": "src",
    "lib": ["es2018"]
  },
  "include": ["packages"]
}
