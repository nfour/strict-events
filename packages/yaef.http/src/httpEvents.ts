import { EventSignature } from 'yaef';

import { IHttpMethod, IStringDict } from './types';

export const HttpRequest = EventSignature('HttpRequest', {} as {
  _eventId: string;
  body: any,
  headers: IStringDict,
  path: string,
  query: IStringDict,
  params: IStringDict,
  method: IHttpMethod,
});

export const HttpRequestResponse = EventSignature('HttpResponse', {} as {
  _eventId: string;
  statusCode: number;
  body: any,
  headers: IStringDict,
});
