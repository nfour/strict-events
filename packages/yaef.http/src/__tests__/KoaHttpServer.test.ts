import { delay } from 'bluebird';
import fetch from 'cross-fetch';
import { ComponentMediator, EventAwaiter } from 'yaef';

import { HttpRequest, HttpRequestResponse } from '../httpEvents';
import {
  AddRouteToKoaHttpServer, KoaHttpServer, KoaHttpServerReady, KoaHttpServerStopped, StartKoaHttpServer,
  StopKoaHttpServer,
} from '../KoaHttpServer';

test('Start a server and send a request and get a response then stop the server', async () => {
  const server = KoaHttpServer({ host: '0.0.0.0', port: 9119 });

  const { connect, mediator } = ComponentMediator({ components: [server] });

  const waitFor = EventAwaiter(mediator);

  await connect();

  // Tell the server to start
  mediator.publish(AddRouteToKoaHttpServer, { methods: ['POST'], path: '/baz' });
  mediator.publish(StartKoaHttpServer);

  await waitFor(KoaHttpServerReady);

  // Hit the server
  const fetchRequestPromise = fetch(`http://0.0.0.0:9119/baz`, {
    method: 'POST',
    headers: { 'content-type': 'application/json', 'accept': 'application/json' },
    body: JSON.stringify({ n: 20 }),
  }).then((res) => res.json());

  mediator.observe(HttpRequest, async (request) => {
    await delay(50); // Take some time to build that response

    expect(request).toMatchSnapshot({
      _eventId: expect.any(String),
    });

    mediator.publish(HttpRequestResponse, {
      body: { nTimes2: request.body.n * 2 },
      headers: { 'content-type': request.headers.accept },
      statusCode: 200,
      _eventId: request._eventId,
    });
  });

  const response = await fetchRequestPromise;

  mediator.publish(StopKoaHttpServer);

  await waitFor(KoaHttpServerStopped);

  expect(response).toMatchSnapshot();
});
