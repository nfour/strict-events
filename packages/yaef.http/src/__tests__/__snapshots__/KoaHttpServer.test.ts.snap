// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Start a server and send a request and get a response 1`] = `
Object {
  "_eventId": Any<String>,
  "body": Object {
    "n": 20,
  },
  "headers": Object {
    "accept": "application/json",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "content-length": "8",
    "content-type": "application/json",
    "host": "0.0.0.0:9999",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
  "method": "POST",
  "name": "HttpRequest",
  "params": Object {},
  "path": "/baz",
  "query": Object {},
}
`;

exports[`Start a server and send a request and get a response 2`] = `
Object {
  "nTimes2": 40,
}
`;
